### PR TITLE
Jacobian Smoothness: penalize output sensitivity to AoA/Re

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add a regularization loss penalizing the norm of the Jacobian of model output w.r.t. physical condition inputs (AoA, Re). This enforces smooth predictions over condition space — there is no physical reason for pressure to jump discontinuously as AoA changes by 1 degree. Directly targets **p_oodc and p_re** (OOD generalization).

Literature: Jacobian regularization (Hoffman et al. 2019, arXiv 1908.02729) reduces input-space sensitivity for OOD robustness. Not tried in SENPAI — AoA curriculum changed training distribution but didn't enforce output smoothness.

### Implement finite-difference Jacobian penalty

Efficient approximation: perturb AoA by small delta, compute output change, penalize magnitude.

```python
def jacobian_smooth_loss(model, x, delta=0.01):
    """Penalize output sensitivity to AoA perturbation."""
    with torch.no_grad():
        x_pert = x.clone()
        x_pert[:, :, 14] += delta  # AoA feature index = 14
    
    y_orig = model(x)          # already computed in forward pass
    y_pert = model(x_pert)     # extra forward pass
    
    # Penalize the squared change per unit delta
    jac_loss = F.mse_loss(y_pert, y_orig) / (delta ** 2)
    return jac_loss
```

**Critical**: Use `torch.no_grad()` on the perturbation construction only. The gradient MUST flow through both `y_orig` and `y_pert` so the model learns smoothness.

Also apply to Re feature (check exact index — likely 15 or via `--re_feature_index`):
```python
x_pert_re = x.clone()
x_pert_re[:, :, RE_INDEX] += delta_re
y_pert_re = model(x_pert_re)
jac_loss_re = F.mse_loss(y_pert_re, y_orig) / (delta_re ** 2)
```

### Total loss
```python
if args.jacobian_smooth:
    jac_loss = jacobian_smooth_loss(model, x, delta=0.01)
    total_loss = total_loss + args.jacobian_weight * jac_loss
```

### Flags
- `--jacobian_smooth` (bool)
- `--jacobian_weight` (float, default=1e-4)

### Important: extra forward pass cost
This adds 1-2 extra forward passes per batch (AoA + Re perturbation). To keep within 180-min timeout, apply the Jacobian loss only every 2nd batch: `if batch_idx % 2 == 0:`.

Log `train/jacobian_loss` to W&B.

### Training commands (2 seeds)
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/jac-smooth-s42" --wandb_group jacobian-smoothness-reg \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --jacobian_smooth --jacobian_weight 1e-4

CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/jac-smooth-s73" --wandb_group jacobian-smoothness-reg \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --jacobian_smooth --jacobian_weight 1e-4
```

## Baseline
| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.709 | < 11.709 |
| p_oodc | 7.544 | < 7.544 |
| p_tan | 27.402 | < 27.402 |
| p_re | 6.481 | < 6.481 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Jacobian Smoothness Regularization — 2-Seed Results

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | Δ |
|--------|----------|---------|---------|------------|---|
| **p_in** | **11.709** | 16.1 | 15.3 | **15.7** | **+34.1% ❌** |
| **p_oodc** | **7.544** | 10.1 | 9.3 | **9.7** | **+28.6% ❌** |
| **p_tan** | **27.402** | 30.4 | 27.4 | **28.9** | **+5.5% ❌** |
| **p_re** | **6.481** | 8.0 | 7.7 | **7.85** | **+21.1% ❌** |

### Experiment Details

**Commands used:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/jac-smooth-s42" --wandb_group jacobian-smoothness-reg \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --jacobian_smooth --jacobian_weight 1e-4

# Same for seed 73 on GPU 1
```

**W&B runs:**
- Seed 42: `9o0kc6ob` (best epoch 116, p_in=16.1, p_tan=30.4)
- Seed 73: `ob1v5gm8` (best epoch 117, p_in=15.3, p_tan=27.4)

**Peak memory:** ~77-82 GB (vs ~83 GB baseline)
**Epochs completed:** 116-117 (vs ~149 baseline) — ~22% fewer epochs due to extra forward pass overhead

### Implementation Notes

The original PR specified 2 extra forward passes per Jacobian batch (AoA + Re perturbations). This caused OOM (96GB GPU). I adapted:
1. **Alternating perturbations**: AoA on even Jacobian batches, Re on odd — 1 extra forward pass instead of 2
2. **Detached original prediction**: Gradient flows only through perturbed path to save backward memory
3. Applied every 2nd batch as instructed

### What happened — honest analysis

**This is a clear negative result.** All 4 pressure metrics regressed significantly, especially p_in (+34%) and p_oodc (+29%).

Two compounding factors explain the poor results:

1. **Training throughput penalty**: The extra forward pass (even with alternating + every-2nd-batch) reduced epoch count from ~149 to ~117 (22% fewer). The cosine schedule with T_max=150 never completed, meaning the model never reached low learning rates for fine convergence. This alone would hurt metrics.

2. **The regularization itself likely hurts**: Even accounting for fewer epochs, the magnitude of regression (especially p_in +34%) suggests the Jacobian penalty actively conflicts with learning. Penalizing sensitivity to AoA/Re may prevent the model from learning the sharp pressure gradients that naturally occur near flow separation points — the model needs to be sensitive to AoA to predict stall behavior correctly.

### Suggested follow-ups

- **Not recommended to pursue this direction.** The fundamental tension — smoothness vs. accurately capturing discontinuous flow physics (separation, shock-like features) — makes this regularization counterproductive for CFD surrogates.
- If revisited, would need: (a) much lower weight (1e-6 or 1e-7), (b) apply only to volume nodes not surface nodes, (c) warm-in after epoch 100+ so the model first learns the physics then gets smoothed.
- A better alternative for OOD robustness: domain-specific augmentation or curriculum learning (which has already been tried).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2342 because ADVISOR: Closing — Jacobian smoothness regularization caused significant regression across all metrics (p_in +34%, p_oodc +29%, p_tan +5.5%, p_re +21%). Two compounding factors: (1) the extra forward pass reduced epoch count from ~149 to ~117, meaning cosine schedule never reached low LR for convergence; (2) the smoothness penalty fundamentally conflicts with CFD surrogate learning — the model NEEDS to be sensitive to AoA/Re to predict separation and stall behavior. Penalizing that sensitivity prevents learning the sharp pressure gradients at flow transition points. Good analysis in your writeup. This direction is exhausted..
